reporting/CassandraBackend.cpp

```diff
@@ -501,68 +501,128 @@ CassandraBackend::fetchBookOffers(
+        std::vector<std::pair<std::uint64_t, ripple::uint256>> keys = {};
+    
+        statement.bindBytes(book.data(), 24);
+        statement.bindInt(seq);
+
+        BOOST_LOG_TRIVIAL(info) << __func__ << " upper = " << std::to_string(seq)
+                                << " book = " << ripple::strHex(std::string((char*)book.data(), 24));
+
+        if (cursor)
+        {
+            auto object = this->fetchLedgerObject(*cursor, seq);
+
+            if(!object)
+                return {{}, {}};
+
+            ripple::SerialIter it{object->data(), object->size()};
+            ripple::SLE offer{it, *cursor};
+            ripple::uint256 bookDir = offer.getFieldH256(ripple::sfBookDirectory);
+
+            statement.bindBytes(bookDir.data() + 24, 8);
+            statement.bindBytes(*cursor);
+        }
         else
         {
-            auto key = result.getUInt256();
-            if (!key.isZero())
-                page.warning = "Data may be incomplete";
+            ripple::uint256 zero = beast::zero;
+            statement.bindBytes(zero.data(), 8);
+            statement.bindBytes(zero);
         }
-    }
-    CassandraStatement statement{selectBook_};
-    statement.bindBytes(book);
-    statement.bindInt(*index);
-    if (cursor)
-        statement.bindBytes(*cursor);
-    else
+
+        // statement.bindUInt(limit);
+        CassandraResult result = this->executeSyncRead(statement);
+
+        BOOST_LOG_TRIVIAL(debug) << __func__ << " - got keys";
+        if (!result)
+        {
+            std::cout << "could not sync read" << std::endl;
+            return {{}, {}};
+        }
+
+        do
+        {
+            auto [quality, index] = result.getBytesTuple();
+            std::uint64_t q = 0;
+            memcpy(&q, quality.data(), 8);
+            keys.push_back({q, ripple::uint256::fromVoid(index.data())});
+
+        } while (result.nextRow());
+
+        return keys;
+    };
+
+    std::vector<std::pair<std::uint64_t, ripple::uint256>> quality_keys;
+    if (sequence != rng->minSequence)
     {
-        statement.bindBytes(zero);
+        upper = (sequence >> 8) << 8;
+        if (upper != sequence)
+            upper += (1 << 8);
+
+        quality_keys = readBooks(book, upper, cursor);
     }
-    statement.bindUInt(limit);
-    CassandraResult result = executeSyncRead(statement);
 
-    BOOST_LOG_TRIVIAL(debug) << __func__ << " - got keys";
-    std::vector<ripple::uint256> keys;
-    if (!result)
-        return page;
-    do
-    {
-        keys.push_back(result.getUInt256());
-    } while (result.nextRow());
-    if (keys.size() && keys.size() == limit)
+    BOOST_LOG_TRIVIAL(debug)
+        << __func__ << " - populated keys. num keys = " << quality_keys.size();
+
+    std::cout << "KEYS SIZE: " << quality_keys.size() << std::endl;
+    if (!quality_keys.size())
+        return {{}, {}};
+
+    std::optional<std::string> warning = {};
+    if (quality_keys[0].second.isZero())
```

Comment:
I don't think this is the correct way to check for completeness. We need to query the zero book to test for completeness, whereas this is looking at the offer keys that correspond to the book passed in as an argument to this function.